File: 2001-3000/2962. Count Subarrays Where Max Element Appears at Least K Times.java
// Medium
// Sliding Window
// Time:O(n), Space:O(1)
// https://leetcode.cn/problems/count-subarrays-where-max-element-appears-at-least-k-times/

class Solution {
    public long countSubarrays(int[] nums, int k) {
        int max = Integer.MIN_VALUE;
        for (int x : nums) {
            max = Math.max(x, max);
        }

        long res = 0;
        int left = 0, cnt = 0;
        for (int right = 0; right < nums.length; right++) {
            if (nums[right] == max) cnt++;
            while (cnt >= k) {
                res += nums.length - right;

                if (nums[left] == max) {
                    cnt--;
                }
                left++;
            }
        }
        return res;
    }
}

File: 2001-3000/2962. Count Subarrays Where Max Element Appears at Least K Times.py
# Time:O(n), Space:O(1)

from typing import List

class Solution:
    def countSubarrays(self, nums: List[int], k: int) -> int:
        max_val = max(nums)
        n = len(nums)

        ans = 0
        cnt_max = 0
        left = 0

        for right in range(n):
            if nums[right] == max_val:
                cnt_max += 1
            
            while cnt_max >= k:
                if nums[left] == max_val:
                    cnt_max -= 1
                left += 1
            
            ans += left
            
        return ans
    